powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java

```diff
@@ -0,0 +1,167 @@
+import org.springframework.stereotype.Controller;
+import org.springframework.validation.BeanPropertyBindingResult;
+import org.springframework.validation.ValidationUtils;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import java.math.BigDecimal;
+import java.security.SecureRandom;
+import java.util.Date;
+import java.util.UUID;
+
+/**
+ * Controller class which handles SMS OTP authorization.
+ *
+ * @author Roman Strobl, roman.strobl@lime-company.eu
+ */
+@Controller
+public class SMSAuthorizationController {
+
+    private SMSAuthorizationRepository smsAuthorizationRepository;
+    private CredentialStoreConfiguration credentialStoreConfiguration;
+
+    public SMSAuthorizationController(SMSAuthorizationRepository smsAuthorizationRepository, CredentialStoreConfiguration credentialStoreConfiguration) {
+        this.smsAuthorizationRepository = smsAuthorizationRepository;
+        this.credentialStoreConfiguration = credentialStoreConfiguration;
+    }
+
+    /**
+     * Create a new SMS OTP authorization message.
+     *
+     * @param request Request data.
+     * @return Response with message ID.
+     */
+    @RequestMapping(value = "/sms/create", method = RequestMethod.POST)
+    public @ResponseBody
+    ObjectResponse<CreateSMSAuthorizationResponse> create(@RequestBody ObjectRequest<CreateSMSAuthorizationRequest> request) throws MethodArgumentNotValidException {
+        CreateSMSAuthorizationRequest createSMSAuthorizationRequest = request.getRequestObject();
+
+        // input validation is handled by CreateSMSAuthorizationRequestValidator
+        // validation is invoked manually because of the generified Request object
+        CreateSMSAuthorizationRequestValidator validator = new CreateSMSAuthorizationRequestValidator();
+        BeanPropertyBindingResult result = new BeanPropertyBindingResult(createSMSAuthorizationRequest, "createSMSAuthorizationRequest");
+        ValidationUtils.invokeValidator(validator, createSMSAuthorizationRequest, result);
+        if (result.hasErrors()) {
+            // getEnclosingMethod() on new object returns a reference to current method
+            MethodParameter methodParam = new MethodParameter(new Object() {
+            }.getClass().getEnclosingMethod(), 0);
+            throw new MethodArgumentNotValidException(methodParam, result);
+        }
+
+        // messageId is generated as random UUID, it can be overriden to provide a real message identification
+        String messageId = UUID.randomUUID().toString();
+
+        // authorization code generation - to be updated based on application requirements
+        String authorizationCode = String.valueOf((new SecureRandom()).nextInt(90000000) + 10000000);
```

Comment:
Implemented as requested, only I switched the key and signatureBase parameters in the hmac.hash() method call. The HMACHashUtilities.hash() method requires a key as the first parameter and data as the second parameter, your sample code has reversed parameters. Please check the last commit, we may need to further tweak the authorization code generation. I commented all lines for this code.